src/project-manager.ts

```diff
@@ -524,42 +674,85 @@ export class InMemoryLanguageServiceHost implements ts.LanguageServiceHost {
 const localFSPlaceholder = 'var dummy_0ff1bd;';
 
 /**
+ * In-memory file cache node which represents either a folder or a file
+ */
+export interface FileSystemNode {
+	file: boolean;
+	children: {[directory: string]: FileSystemNode};
+};
+
+/**
  * In-memory file system, can be served as a ParseConfigHost (thus allowing listing files that belong to project based on tsconfig.json options)
  */
 export class InMemoryFileSystem implements ts.ParseConfigHost, ts.ModuleResolutionHost {
 
-	entries: any;
-	overlay: any;
+	/**
+	 * Map (filepath -> string content) of files fetched from the remote file system. Paths are relative to `this.path`
+	 */
+	entries: { [filePath: string]: string };
+
+	/**
+	 * Map (filepath -> string content) of temporary files made while user modifies local file(s).  Paths are relative to `this.path`
+	 */
+	overlay: { [filePath: string]: string };
+
+	/**
+	 * Should we take into account register when performing a file name match or not. On Windows when using local file system, file names are case-insensitive
+	 */
 	useCaseSensitiveFileNames: boolean;
+
+	/**
+	 * Root path
+	 */
 	path: string;
 
-	rootNode: any;
+	/**
+	 * File tree root
+	 */
+	rootNode: FileSystemNode;
 
 	constructor(path: string) {
 		this.path = path;
 		this.entries = {};
 		this.overlay = {};
-		this.rootNode = {};
+		this.rootNode = {file: false, children: {}};
```

Comment:
This will cause errors for folders or files that are named after a built in `Object` property, for example `hasOwnProperty`, `valueOf`, `toString`. `children` must be `Object.create(null)`

Can do this in a follow-up PR though.